File: src/main/java/fortress/tfol/visitor/TermVisitor.java
package fortress.tfol.visitor;

import fortress.tfol.*;

public interface TermVisitor<T> {
    default public T visit(Term term) {
        return term.accept(this);
    }
    
    public T visitTop(Top term);
    public T visitBottom(Bottom term);
    public T visitVar(Var term);
    public T visitNot(Not term);
    public T visitAndList(AndList term);
    public T visitOrList(OrList term);
    public T visitDistinct(Distinct term);
    public T visitIff(Iff term);
    public T visitImplication(Implication term);
    public T visitEq(Eq term);
    public T visitApp(App term);
    public T visitExists(Exists term);
    public T visitForall(Forall term);
}

File: core/src/main/scala/fortress/msfol/TermVisitor.scala
package fortress.msfol

import fortress.msfol._

/** A visitor for term syntax trees. */
trait TermVisitor[T] {
    def visit(term: Term): T = term.accept(this)

    def visitTop(): T
    def visitBottom(): T
    def visitVar(term: Var): T
    def visitEnumValue(term: EnumValue): T
    def visitDomainElement(term: DomainElement): T
    def visitNot(term: Not): T
    def visitAndList(term: AndList): T
    def visitOrList(term: OrList): T
    def visitDistinct(term: Distinct): T
    def visitImplication(term: Implication): T
    def visitIff(term: Iff): T
    def visitEq(term: Eq): T
    def visitApp(term: App): T
    def visitBuiltinApp(term: BuiltinApp): T
    def visitExists(term: Exists): T
    def visitForall(term: Forall): T
    def visitIntegerLiteral(term: IntegerLiteral): T
    def visitBitVectorLiteral(term: BitVectorLiteral): T
    def visitIfThenElse(term: IfThenElse): T
    def visitClosure(term: Closure): T
    def visitReflexiveClosure(term: ReflexiveClosure): T
}
